### PR TITLE
Allow adding new topic

### PR DESCRIPTION
**Pull request by SourabhTestpress**
Added CreateTopicView to allow logged in user add a new topic.

Stats

Number of Test Cases - 4
Number of SQL Queries - 3

Guidelines

PEP8(Python)/Testpress JS compatible Yes
Is each logical block separated with a line break? Yes
Have you self reviewed this PR in context to the previous PR feedbacks? Yes

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -15,27 +15,23 @@ jobs:
...
     machine:
       enabled: true
     steps:
       - checkout
       - run:
            name: Deploy Master to Heroku
            command: |
-             git push https://heroku:$HEROKU_API_KEY@git.heroku.com/$HEROKU_APP_NAME.git master
-
-
```
Comment (SourabhTestpress) [inline .circleci/config.yml:33]: In master branch also these changes are present but it is still showing these changes, I can't understand why?
https://github.com/SourabhTestpress/NewBoardApp/blob/master/.circleci/config.yml

Comment (waranlogesh) [reply .circleci/config.yml:33]: Rebase this branch properly. Look into `git rebase -i`

File: boards/tests/test_views.py [hunk 2]
```diff
@@ -58,3 +60,49 @@ def test_board_detail_view_produces_status_code_200(self):
         request = self.factory.get(reverse("detail", args=(self.board_obj.slug,)))
         response = board_detail(request, self.board_obj.slug)
         self.assertEquals(200, response.status_code)
+
+
+def create_a_user_and_a_board():
```
Comment (waranlogesh) [inline boards/tests/test_views.py:65]: This violates SRP

Comment (waranlogesh) [inline boards/tests/test_views.py:65]: This violates SRP

File: boards/tests/test_views.py [hunk 3]
```diff
@@ -58,3 +60,49 @@ def test_board_detail_view_produces_status_code_200(self):
...
+        request = self.factory.get(reverse("new_topic", args=(self.board_obj.slug,)))
+        request.user = self.user
+        response = CreateTopicView.as_view()(request, self.board_obj.slug)
+        self.assertContains(response, "csrfmiddlewaretoken")
+
+    def test_CreateTopicView_redirects_on_valid_post_data(self):
+        url = reverse("new_topic", args=(self.board_obj.slug,))
+        data = {"subject": "Test title", "message": "Lorem ipsum dolor sit amet"}
+        self.client.force_login(self.user)
+        response = self.client.post(url, data)
```
Comment (waranlogesh) [inline boards/tests/test_views.py:95]: Don't use client.

File: boards/tests/test_views.py [hunk 4]
```diff
@@ -58,3 +60,49 @@ def test_board_detail_view_produces_status_code_200(self):
...
+        self.assertRedirects(response, reverse("detail", args=(self.board_obj.slug,)))
+
+    def test_should_render_error_on_no_form_data(self):
+        """
+        We are sending an empty dictionary.
```
Comment (waranlogesh) [inline boards/tests/test_views.py:100]: Comments not needed

File: boards/tests/test_views.py [hunk 5]
```diff
@@ -58,3 +60,49 @@ def test_board_detail_view_produces_status_code_200(self):
...
+        The expected behavior is to show the form again with validation errors
+        """
+        url = reverse("new_topic", args=(self.board_obj.slug,))
+        data = {}
+        request = self.factory.post(url, data)
+        request.user = self.user
+        response = CreateTopicView.as_view()(request, self.board_obj.slug)
+        self.assertEquals(response.status_code, 200)
```
Comment (waranlogesh) [inline boards/tests/test_views.py:108]: Where are we testing whether error is rendered or not?

File: boards/views.py [hunk 6]
```diff
@@ -26,3 +31,25 @@ def topic_detail(request, slug):
...
     context = {"topic": topic}
     return render(request, "topic/detail.html", context)
+
+
+class CreateTopicView(LoginRequiredMixin, CreateView):
+    model = Topic
+    form_class = CreateTopicForm
+    template_name = "topic/new_topic.html"
+
+    def form_valid(self, form):
```
Comment (waranlogesh) [inline boards/views.py:41]: Lots of code smells in this method.

**Review by waranlogesh — CHANGES_REQUESTED** [review]

File: boards/tests/test_views.py [hunk 7]
```diff
@@ -58,3 +60,48 @@ def test_board_detail_view_produces_status_code_200(self):
...
+        self.assertTrue(Topic.objects.exists())
+        self.assertTrue(Post.objects.exists())
+
+    def test_should_render_error_on_no_form_data(self):
+        url = reverse("new_topic", kwargs={"slug": self.board_obj.slug})
+        data = {}
+        self.client.force_login(self.user)
+        response = self.client.post(url, data)
+        form = response.context.get("form")
+        self.assertTrue(form.errors)
```
Comment (SourabhTestpress) [inline boards/tests/test_views.py:107]: To test context data, we have to go through middlewares,so I used client. This is according to this [stackoverflow answer](https://stackoverflow.com/a/11553035/14354318), when I tried to use it using request factory I got error "Template response has no attribute context"

File: boards/tests/test_views.py [hunk 8]
```diff
@@ -58,3 +60,48 @@ def test_board_detail_view_produces_status_code_200(self):
...
+    def setUp(self):
+        self.board_obj = Board.objects.create(
+            name="board", slug="boarddslug", description="1"
+        )
+        self.user = User.objects.create_user(
+            username="sourabh", email="so@dj.com", password="Somepassword"
+        )
+        self.factory = RequestFactory()
+
+    def test_csrf_token_is_present_in_get_request_to_CreateTopicView(self):
```
Comment (waranlogesh) [inline boards/tests/test_views.py:75]: Follow PEP8 method naming conventions.

File: boards/tests/test_views.py (hunk 7 above)
Comment (waranlogesh) [reply boards/tests/test_views.py:107]: Then this shouldn't be tested at view level. Add test case for the form.

File: boards/views.py (hunk 6 above)
Comment (waranlogesh) [reply boards/views.py:41]: Move this to  form's save(). View should be thin.

**Review by waranlogesh — CHANGES_REQUESTED** [review]

File: boards/forms.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+class CreateTopicForm(forms.ModelForm):
+    message = forms.CharField(widget=forms.Textarea(), max_length=2000)
+
+    class Meta:
+        model = Topic
+        fields = ["subject", "message"]
+
+    def save(self, current_user, **kwargs):
+        self.instance.board = get_object_or_404(Board, slug=kwargs.get("slug"))
```
Comment (waranlogesh) [inline boards/forms.py:16]: Board should be added in form as a hidden field and auto populated.

File: boards/forms.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+        self.instance.user = current_user
```
Comment (waranlogesh) [inline boards/forms.py:17]: Same add a user as a hidden form field and auto-populate.

File: boards/tests/test_forms.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        )
+        self.user = User.objects.create_user(
+            username="sourabh", email="so@dj.com", password="Somepassword"
+        )
+
+    def test_should_render_error_on_no_form_data(self):
+        url = reverse("new_topic", kwargs={"slug": self.board_obj.slug})
+        data = {}
+        self.client.force_login(self.user)
+        response = self.client.post(url, data)
```
Comment (waranlogesh) [inline boards/tests/test_forms.py:21]: Import the form and test it directly. No need to go through client for checking form errors.

**Review by waranlogesh — CHANGES_REQUESTED** [review]

File: boards/tests/test_forms.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+from django.urls import reverse
+from django.test import SimpleTestCase, TestCase
+from django.contrib.auth.models import User
+
+from boards.forms import CreateTopicForm
+from boards.models import Board
+
+
+class AddNewTopicTest(TestCase):
```
Comment (waranlogesh) [inline boards/tests/test_forms.py:9]: This test case is not touching DB so it should be SimpleTestCase

File: boards/tests/test_forms.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+from django.urls import reverse
+from django.test import SimpleTestCase, TestCase
```
Comment (waranlogesh) [inline boards/tests/test_forms.py:2]: Isn't flake8 configured as a pre-commit hook? Unused imports.

**Review by waranlogesh — CHANGES_REQUESTED** [review]

**Comment by SourabhTestpress** [comment]
Leaving a test comment

**Pull request merged**